libherokubuildpack/src/buildpack_output/mod.rs

```diff
@@ -502,6 +569,30 @@ where
     }
 }
 
+impl<W> BuildpackOutput<state::Background<W>>
+where
+    W: Write + Send + Sync + 'static,
+{
+    /// Finalize a timer's output
+    ///
+    /// Once you're finished with your long running task, calling this function
+    /// finalizes the timer's output and transitions back to a [`state::Section`].
+    #[must_use]
+    pub fn finish(self) -> BuildpackOutput<state::Section<W>> {
+        let duration = self.state.started.elapsed();
+        let mut io = match self.state.write.stop() {
+            Ok(io) => io,
+            Err(e) => std::panic::resume_unwind(e),
+        };
```

Comment:
This match and resuming of the panic was suggested by clippy for handling a thread result. The alternative would be manually panicking, which...either way, is a panic.

Reply:
Why not use `let mut io = self.state.write.stop().expect("...");` instead and provide a message stating that the internal restrictions of the `PrintGuard` have been violated?

Reply:
I'm not 100% sure, but this is the recommendation from stdlib
https://doc.rust-lang.org/std/thread/type.Result.html

> The value contained in the Result::Err variant is the value the thread panicked with; that is, the argument the panic! macro was called with. Unlike with normal errors, this value doesn’t implement the [Error](https://doc.rust-lang.org/std/error/trait.Error.html) trait.
>
> Thus, a sensible way to handle a thread panic is to either:
>
> propagate the panic with [std::panic::resume_unwind](https://doc.rust-lang.org/std/panic/fn.resume_unwind.html)

If we add an `.expect()` then clippy complains that this function doesn't have a `# Panics` section, but with this code recommended by stdlib docs, it doesn't. I figured it was better to go with the recommended version that requires fewer docs or lint disables to make clippy happy.

If it's eyebrow raising we could add a comment linking to those docs.

Reply:
Maybe, for consistency, the same should apply to the following code in this PR?
https://github.com/heroku/libcnb.rs/blob/841817f5d2645cfefcc3f2e5b6aef174be0cac91/libherokubuildpack/src/buildpack_output/mod.rs#L538-L545

Reply:
> Maybe, for consistency, the same should apply to the following code in this PR?

I went ahead and annotated my code with a comment from my prior musing:

> If it's eyebrow raising we could add a comment linking to those docs.

I think that's good enough. 

In this case we're handling a different class of error. This code is otherwise identical to `writeln_now` which writes and flushes, both can panic if the writer is closed. In that case I believe we want to add an annotation here otherwise the panic wouldn't have the information that it was caused while we were trying to write to the buildpack output.

Contrast this with the panic resume_unwind. In that case either the problem happened in a different thread and has one of two causes (if I understand correctly). 

- Our code panic-d and already added an expectation message https://github.com/heroku/libcnb.rs/pull/784/files#diff-bacb3b82540d939a8804ab0a93bb0cc89eb3c098f073a0036f77dcbe660f17ffR29
- Or some unknown system level event, like an OS bug, triggered a panic in the thread, in which case we don't want to annotate the error on join, we want to effectively continue it.

I'm not 100% sure of all the cases when that second one can be triggered, but even if we ignore it, i think it makes sense to propagate the panic in the background thread case, but write an `except` message here.